lib/db/backend/badger_backend.go

```diff
@@ -214,22 +176,33 @@ func (b *badgerBackend) Delete(key []byte) error {
 }
 
 func (b *badgerBackend) Compact() error {
-	b.closedMut.RLock()
-	defer b.closedMut.RUnlock()
-	if b.closed {
-		return errClosed{}
+	if err := b.closeWG.Add(1); err != nil {
+		return err
 	}
-
-	b.closeWG.Add(1)
 	defer b.closeWG.Done()
 
-	// XXX: check if this is appropriate or if we also need Flatten and
-	// whatnot.
-	err := b.bdb.RunValueLogGC(0.5)
+	// This weird looking loop is as recommended in the README
+	// (https://github.com/dgraph-io/badger/tree/v2.0.3#garbage-collection).
+	// Basically, the RunValueLogGC will pick some promising thing to
+	// garbage collect at random and return nil if it improved the
+	// situation, then return ErrNoRewrite when there is nothing more to GC.
+	// The 0.5 is the discard ratio, for which the method docs say they
+	// "recommend setting discardRatio to 0.5, thus indicating that a file
+	// be rewritten if half the space can be discarded".
+	var err error
+	for err == nil {
+		err = b.bdb.RunValueLogGC(0.5)
+	}
```

Comment:
I am not expecting this to fail, but if badger ever does get into a weird state where it returns `nil` forever, this is going to be a silent complete hangup. Maybe worth adding a timer into the mix and info log every 10 min (or whatever magic number comes to mind) that compaction is still ongoing - that would then at least point at what the problem is. I wouldn't do a hard limit on time, there's some seriously slow devices out there in "normal use".